File: backend/app/services/sheets.py
"""
Answer-sheet storage and ingestion.

Two jobs: keep uploaded sheets in a private bucket so a teacher can look at the
paper while reviewing a mark, and turn whatever the scanner produced into one
sheet per student — including the common case where the whole class arrives as
a single multi-page PDF.
"""

from __future__ import annotations

import io
import re
import unicodedata
from dataclasses import dataclass

from fastapi import HTTPException, UploadFile
from supabase import Client

BUCKET = "answer-sheets"
MAX_BYTES = 25 * 1024 * 1024
ALLOWED = {"image/jpeg", "image/png", "image/webp", "image/heic", "application/pdf"}


@dataclass
class SheetFile:
    """One student's sheet, ready to store."""

    file_name: str
    mime_type: str
    content: bytes
    page_from: int | None = None
    page_to: int | None = None


class SheetStorage:
    def __init__(self, db: Client) -> None:
        self.db = db

    def _bucket(self):
        return self.db.storage.from_(BUCKET)

    def upload(self, owner_id: str, test_id: str, student_id: str, sheet: SheetFile) -> str:
        path = f"{owner_id}/{test_id}/{student_id}-{_safe_name(sheet.file_name)}"
        self._bucket().upload(
            path,
            sheet.content,
            {"content-type": sheet.mime_type, "upsert": "true"},
        )
        return path

    def download(self, storage_path: str) -> bytes:
        return self._bucket().download(storage_path)

    def delete_many(self, paths: list[str]) -> None:
        cleaned = [path for path in paths if path]
        if cleaned:
            self._bucket().remove(cleaned)


def _safe_name(name: str) -> str:
    normalized = unicodedata.normalize("NFKD", name).encode("ascii", "ignore").decode()
    cleaned = re.sub(r"[^A-Za-z0-9._-]+", "-", normalized).strip("-")
    return (cleaned or "sheet")[:80]


async def read_upload(file: UploadFile) -> tuple[bytes, str]:
    content = await file.read()
    if not content:
        raise HTTPException(status_code=400, detail=f"{file.filename} is empty")
    if len(content) > MAX_BYTES:
        raise HTTPException(status_code=413, detail=f"{file.filename} is larger than 25MB")
    mime = (file.content_type or "").lower()
    if mime not in ALLOWED:
        raise HTTPException(
            status_code=415,
            detail=f"{file.filename} is a {mime or 'unknown'} file. Upload images or PDFs.",
        )
    return content, mime


def pdf_page_count(content: bytes) -> int:
    from pypdf import PdfReader

    return len(PdfReader(io.BytesIO(content)).pages)


def extract_pdf_pages(content: bytes, first: int, last: int) -> bytes:
    """Slice an inclusive, 1-based page range into a standalone PDF."""
    from pypdf import PdfReader, PdfWriter

    reader = PdfReader(io.BytesIO(content))
    writer = PdfWriter()
    for index in range(first - 1, min(last, len(reader.pages))):
        writer.add_page(reader.pages[index])
    buffer = io.BytesIO()
    writer.write(buffer)
    return buffer.getvalue()


def group_pages_by_student(names: list[str | None]) -> list[tuple[int, int, str | None]]:
    """
    Turn a per-page list of detected names into contiguous per-student ranges.

    A page with no readable name continues the previous student, which is what a
    multi-page answer booklet looks like: the name is on the front sheet only.
    Returns 1-based inclusive `(first_page, last_page, name)` tuples.
    """
    groups: list[tuple[int, int, str | None]] = []
    for index, name in enumerate(names, start=1):
        normalized = (name or "").strip() or None
        if groups and (normalized is None or normalized == groups[-1][2]):
            first, _, current = groups[-1]
            groups[-1] = (first, index, current or normalized)
        else:
            groups.append((index, index, normalized))
    return groups


def match_name_to_student(name: str | None, students: list[dict]) -> str | None:
    """Best-effort name → student id. Exact, then surname/first-name overlap."""
    if not name:
        return None
    target = _normalize(name)
    if not target:
        return None

    for student in students:
        if _normalize(student["name"]) == target:
            return student["id"]

    target_parts = set(target.split())
    best: tuple[int, str] | None = None
    for student in students:
        parts = set(_normalize(student["name"]).split())
        overlap = len(target_parts & parts)
        if overlap and (best is None or overlap > best[0]):
            best = (overlap, student["id"])
    return best[1] if best else None


def _normalize(value: str) -> str:
    return re.sub(r"[^a-z0-9 ]+", " ", value.lower()).strip()

File: src/lib/server/sheets.js
/**
 * Turning whatever the scanner produced into one sheet per student.
 *
 * The common case is not one file per student — it is the whole class in a
 * single multi-page PDF, which has to be split by reading the name off each
 * page.
 */
import { PDFDocument } from "pdf-lib";
import { ApiError } from "./http";

async function loadPdf(content) {
  try {
    return await PDFDocument.load(content, { ignoreEncryption: true });
  } catch {
    // A truncated download or a mislabeled file is the uploader's problem to
    // hear about, not a server error.
    throw new ApiError(422, "That PDF could not be read. Re-export or re-scan it and try again.");
  }
}
export async function pdfPageCount(content) {
  return (await loadPdf(content)).getPageCount();
}

/** Slice an inclusive, 1-based page range into a standalone PDF. */
export async function extractPdfPages(content, first, last) {
  const source = await loadPdf(content);
  const target = await PDFDocument.create();
  const total = source.getPageCount();
  const indices = [];
  for (let index = first - 1; index < Math.min(last, total); index += 1) indices.push(index);
  const pages = await target.copyPages(source, indices);
  pages.forEach((page) => target.addPage(page));
  return target.save();
}

/**
 * Turn a per-page list of detected names into contiguous per-student ranges.
 *
 * A page with no readable name continues the previous student, which is what a
 * multi-page answer booklet looks like: the name is on the front sheet only.
 * Returns 1-based inclusive ranges.
 */
export function groupPagesByStudent(names) {
  const groups = [];
  names.forEach((raw, index) => {
    const page = index + 1;
    const name = (raw || "").trim() || null;
    const previous = groups[groups.length - 1];
    if (previous && (name === null || name === previous.name)) {
      previous.last = page;
      previous.name = previous.name || name;
    } else {
      groups.push({
        first: page,
        last: page,
        name,
      });
    }
  });
  return groups;
}
/** Best-effort name → student. Exact match first, then first/surname overlap. */
export function matchNameToStudent(name, students) {
  if (!name) return null;
  const target = normalize(name);
  if (!target) return null;
  for (const student of students) {
    if (normalize(student.name) === target) return student.id;
  }
  const wanted = new Set(target.split(" ").filter(Boolean));
  let best = null;
  for (const student of students) {
    const parts = new Set(normalize(student.name).split(" ").filter(Boolean));
    let overlap = 0;
    wanted.forEach((part) => {
      if (parts.has(part)) overlap += 1;
    });
    if (overlap && (!best || overlap > best.overlap))
      best = {
        overlap,
        id: student.id,
      };
  }
  return best?.id ?? null;
}

/**
 * Match one uploaded file to a student by its name alone.
 *
 * Student code first, then every name part appearing in the filename, then a
 * roll number. Anything ambiguous returns null — see the note in the upload
 * route about why a wrong match is worse than no match.
 */
export function matchByFilename(fileName, students, taken) {
  const stem = normalize(fileName.replace(/\.[^.]+$/, ""));
  const free = students.filter((student) => !taken.has(student.id));
  for (const student of free) {
    if (student.code && stem.includes(normalize(student.code))) return student.id;
  }
  for (const student of free) {
    const parts = normalize(student.name)
      .split(" ")
      .filter((part) => part.length > 2);
    if (parts.length && parts.every((part) => stem.includes(part))) return student.id;
  }
  for (const value of stem.match(/\b\d{1,4}\b/g) ?? []) {
    for (const student of free) {
      if (student.roll_no && String(Number(value)) === student.roll_no.trim()) return student.id;
    }
  }
  return null;
}
function normalize(value) {
  return value
    .toLowerCase()
    .replace(/[^a-z0-9 ]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}
